middlewares/statsd.go

```diff
@@ -0,0 +1,85 @@
+package middlewares
+
+import (
+	"time"
+
+	"github.com/containous/traefik/log"
+	"github.com/containous/traefik/safe"
+	"github.com/containous/traefik/types"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/metrics"
+	"github.com/go-kit/kit/metrics/statsd"
+)
+
+var _ Metrics = (Metrics)(nil)
+
+var statsdClient = statsd.New("traefik.", kitlog.LoggerFunc(func(keyvals ...interface{}) error {
+	log.Info(keyvals)
+	return nil
+}))
+var statsdTicker *time.Ticker
+
+// Statsd is an Implementation for Metrics that exposes statsd metrics for the latency
+// and the number of requests partitioned by status code and method.
+// - number of requests partitioned by status code and method
+// - request durations
+// - amount of retries happened
+type Statsd struct {
+	reqsCounter          metrics.Counter
+	reqDurationHistogram metrics.Histogram
+	retryCounter         metrics.Counter
+}
+
+func (s *Statsd) getReqsCounter() metrics.Counter {
+	return s.reqsCounter
+}
+
+func (s *Statsd) getReqDurationHistogram() metrics.Histogram {
+	return s.reqDurationHistogram
+}
+
+func (s *Statsd) getRetryCounter() metrics.Counter {
+	return s.retryCounter
+}
+
+// NewStatsD creates new instance of StatsD
+func NewStatsD(name string) *Statsd {
+	var m Statsd
+
+	m.reqsCounter = statsdClient.NewCounter(ddMetricsReqsName, 1.0).With("service", name)
+	m.reqDurationHistogram = statsdClient.NewTiming(ddMetricsLatencyName, 1.0).With("service", name)
+
+	return &m
+}
+
+// InitStatsdClient initializes metrics pusher and creates a statsdClient if not created already
+func InitStatsdClient(config *types.Statsd) *time.Ticker {
```

Comment:
So I looked once again at the `InitDataDogClient` and `InitStatsDClient` code and the only part they share is this:
```
                address := config.Address
		if len(address) == 0 {
			address = "localhost:8125"
		}
		pushInterval, err := time.ParseDuration(config.PushInterval)
		if err != nil {
			log.Warnf("Unable to parse %s into pushInterval", config.PushInterval)
			pushInterval = 10 * time.Second
		}
```
Given that it looks like boiler-plate code, it's really just a validation of `config` properties (but keep in mind that `config` is of different struct type: `types.Datadog` vs. `types.Statsd`), they just happen to both have `address` and `pushInterval` properties).
The rest of the function handles initializing different clients and each provider has their own ticker as well, which is a statically managed variable, so not that great for factorizing.  IMHO it would be more code and will make it harder to read and reason about the code if we to try to come up with a base config type (not even sure what to call it) and then have the 2 concrete types encapsulate it.  While I do agree that on the surface this does seem like a good possible case for boiler-plate reduction, once you look deeper, that does not turn out to be so, but curious to hear what @emilevauge thinks about this.